library/std/src/sys/pal/windows/fs.rs

```diff
@@ -1068,14 +1068,35 @@ pub fn readdir(p: &Path) -> io::Result<ReadDir> {
     unsafe {
         let mut wfd = mem::zeroed();
         let find_handle = c::FindFirstFileW(path.as_ptr(), &mut wfd);
+
+        // The status `ERROR_FILE_NOT_FOUND` is returned by the `FindFirstFileW` function
+        // if no matching files can be found, but not necessarily that the path to find the
+        // files in does not exist.
+        //
+        // Hence, a check for whether the path to search in exists is added when the last
+        // os error returned by Windows is `ERROR_FILE_NOT_FOUND` to handle this scenario.
+        // If that is the case, an empty `ReadDir` iterator is returned as it returns `None`
+        // in the initial `.next()` invocation because `ERROR_NO_MORE_FILES` would have been
+        // returned by the `FindNextFileW` function.
+        //
+        // See issue #120040: https://github.com/rust-lang/rust/issues/120040.
+        let last_error = Error::last_os_error();
+        if last_error.raw_os_error().unwrap() == c::ERROR_FILE_NOT_FOUND as i32 && p.exists() {
```

Comment:
I'm not convinced `p.exists()` is a good idea. Firstly it's a [time-of-check to time-of-use](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) hazard. The path may have been removed after the `FindFirstFileW` call but before the `p.exists` check. Secondly, if it didn't exist then `FindFirstFileW` would have failed with `ERROR_PATH_NOT_FOUND` instead of `ERROR_FILE_NOT_FOUND`.